tinygrad/codegen/linearizer.py

```diff
@@ -180,13 +180,21 @@ def render_loop(self, xx:List[Variable], depth:int) -> Tuple[UOp, ...]:
     self.loop_uops.update(new_loops)
     return tuple(new_loops.values())
 
-  def render_reduceop(self, reduceop:LazyOp, accs:Dict[LazyOp, List[UOp]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]],
-                      global_idxs, local_idxs, upcast_idxs):
+  # helper function, the actual render_reduceop function will get initialize once the indexes get initilized
+  def render_reduceop(self, reduceop:LazyOp, seq: int, accs:Dict[LazyOp, List[UOp]],
+                       loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]],
+                      global_idxs, local_idxs, reduce_idxs, upcast_idxs, full_upcast_idxs):
     # define indicies
-    full_upcast_idxs = [Variable(f"_uidx{i}", 0, s-1) for i, s in enumerate(self.full_shape[self.shape_len-self.upcasted:])]
-    reduce_idxs = [Variable(f"ridx{i}", 0, self.full_shape[i]-1) for i in range(self.first_reduce+self.group_for_reduces, self.shape_len-self.upcasted)]  # noqa: E501
     fake_reduce_idxs = [x*0 for x in reduce_idxs]
 
+    # add a local buffer for multistage reduce. # TODO: use local alias
```

Comment:
I need to take more time with this part - Can you explain why this was moved from https://github.com/tinygrad/tinygrad/pull/4259/files#diff-63d57daaca59e5495bbb108da1c13e782f4a408db8503e015e8094ac07c87ea6L354

Reply:
yeah so for sequential reduceops you should be able to just reuse the same buffer which is why for a while this block of code was in the `linearize` function

however, it seems to miss behave on GPU: ex the CI for [this commit](https://github.com/tinygrad/tinygrad/pull/4259/commits/c10986b648787bf1c0753882d7152ca97264b1f7) versus the next (which actually makes each local buffer unique)

I'm honestly not sure why, because I can't get the bug to reproduce on my machine or the tinybox but I could trace the issue to the local buffers and giving each reduceop it's own buffer fixed it

I don't think it has anything to do with the `earlybufs`

Reply:
so that it gets run for every reduceop

we could keep it in 'linearize' and give it it's own for loop but it seemed more natural to put it in 'render_reduceop' so we can keep using '-1' to reference the local_buffer in stuff like 'self.sts' or 'global_store'

Reply:
I think moving this codeblock to render_reduceop can be its own noop change, makes this diff more readable.